presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1250,6 +1253,7 @@ public PlanWithProperties visitUnion(UnionNode node, PreferredProperties parentP
                                                 source.getNode().getOutputVariables(),
                                                 Optional.empty(),
                                                 nullsAndAnyReplicated,
+                                                COLUMNAR,
```

Comment:
At the step when exchanges are added the optimizer sets the encoding to COLUMNAR. There's a rule downstream checking what would be the optimal encoding and switching to ROW_WISE is necessary. 

At the point when exchanges are added the number of columns exchanged may not be final, hence it might be too early to pick an optimal one at this step.